Refactor to expose experimental hook for overriding prettier options (#259)
ember-template-lint/ember-template-lint-plugin-prettier: README.md

The recommended set will apply [these rules](https://github.com/ember-template-lint/ember-template-lint-plugin-prettier/blob/v1.1.0-beta.0/lib/config/recommended.js).

> ## Configure Prettier Options
> 
> Prettier can only be configured via [standard prettier config files](https://prettier.io/docs/en/configuration.html)
> It is not possible to pass prettier options with the rule options.

Reviewer: ```suggestion
## Configuration

Prettier can be configured via [standard prettier config files](https://prettier.io/docs/en/configuration.html).
```

The recommended set will apply [these rules](https://github.com/ember-template-lint/ember-template-lint-plugin-prettier/blob/v1.1.0-beta.0/lib/config/recommended.js).

## Configure Prettier Options

> Prettier can be configured via prettier rc file and rule config.

Reviewer: ```suggestion
Prettier can be configured via [standard prettier config files](https://prettier.io/docs/en/configuration.html), or rule config options.
```

> Prettier can be configured via [standard prettier config files](https://prettier.io/docs/en/configuration.html), or rule config options.

Reviewer: Another open question, is it a good idea to support inline config of prettier at all? Prettier config passed directly to ember-template-lint will not be recognized when prettier is run directly (`prettier file.hbs`) or from a prettier IDE extension. That seems like a good reason to not support inline prettier config. People will be confused when these prettier options apply when they run it from ember-template-lint but not from elsewhere. So I think I'd rather require that prettier is configured using standard prettier config files, and that we don't create our own non-standard, copycat version of prettier config.

CC: @dcyriller

Author: eslint prettier plugin also supports it. That's ehy i was expecting it to work here to

Author: https://github.com/prettier/eslint-plugin-prettier#options

Reviewer 2: Thank you @patricklx for taking the time to contribute!

I would not recommend specifying the prettier options in the ember-template-lint config. This can lead to inconsistencies in code formatting between editors or when running the `npm run ember-template-lint .` command. Editors will not read the prettier config from the ember-template-lint config file, but from the prettier config file (.prettierrc.js, for instance).

Instead, I recommend the following approach. This is the setup I've been using for a while. To be honest, I no longer use the ember-template-lint-plugin-prettier library. Instead, I use the following package.json scripts to lint and format:
```
 "scripts": {
     "format": "prettier . --write",
     "lint": "npm-run-all --parallel lint:*",
     "lint:eslint": "eslint .",
     "lint:prettier": "prettier . --check",
     "lint:template-lint": "ember-template-lint .",
     ...
   },
```
Note: npm-run-all needs to be installed in the `devDependencies`.

This way, each tool has its own config and its own script. I find it simpler to use, understand, and maintain.

Author: okay, but then i would like to document it at least. That this reads the prettier configuration from the prettierrc file. 
 so you are using prettier directly for hbs files? Gts is still not supported... But works in ember-template-lint.
Is it possible for prettier to have different options based on file type/extension?

Author: Btw, i created the eslint plugin for template lint, which is why i also wanted to add this option. To have it work the same as the prettier eslint plugin

Reviewer 2: > okay, but then i would like to document it at least

Sounds great, :+1:

> so you are using prettier directly for hbs files?

Yes, I do. In CI, I run the `npm run lint` script which executes `npm run lint:prettier` (among other linters).

> Btw, i created the eslint plugin for template lint, which is why i also wanted to add this option. To have it work the same as the prettier eslint plugin

I wasn't aware of the library. Thank you for providing me with this context. I have updated my previous suggestion so that it works in the context of eslint-plugin-ember-template-lint.

Author: i changed the code so that it uses a function call to obtain the prettier options, with that I can hook into in the eslint-plugin and return custom options

Author: @dcyriller @bmish

Reviewer: I think moving the prettier config retrieval logic into a separate function sounds good. That way, there's no behavior change, and you can just override that function when you extend from this rule, right? Can you update the PR title and description to explain this and provide some context?

The configuration from the rule will be merged with the rc file config, overriding existing options.
e.g

```js
module.exports = {
  plugins: ["ember-template-lint-plugin-prettier"],

  extends: ["recommended", "ember-template-lint-plugin-prettier:recommended"],
  
  rules: {

>     "prettier": { singleQuote: true }

Author: We could also check if the config is an array, take the first as prettier options and others for the future. Like in eslint

>     "prettier": { singleQuote: true }

Author: I was thinking to support it like this:
`prettier: ['error', {...prettierConf}, {...otherConf}]`
actually not possible, as template-lint only passes the first option: https://github.com/ember-template-lint/ember-template-lint/blob/master/lib/helpers/determine-rule-config.js#L35
another options that can also done:
`prettier: ['error', [{...prettierConf}, {...otherConf}]]`

>     "prettier": { singleQuote: true }

Reviewer: We decided against this option in the other thread: https://github.com/ember-template-lint/ember-template-lint-plugin-prettier/pull/259#discussion_r1210607256
